### DISCUSSION TITLE
saving data with on_close doesn't work after certain point in experiment

### DISCUSSION THREAD

**Discussion by patiencerosestevens** [jspsych/jsPsych#1596] [Q&A] [answered]
Hello, I'm having trouble with saving data when participants close the browser before the end of the experiment. I'm saving data for both on_close and on_finish  when I run the experiment timeline (reached_finish is set to 0 at beginning of expt) : 

```
jsPsych.init({
timeline: timeline,
on_close: function () {
if (reached_finish == 0) {
    jsPsych.data.addProperties({
          'subject': subjID,
          'assignedYesKey': jsPsych.pluginAPI.convertKeyCharacterToKeyCode(yeskey),
          'wordList': rand_stim_sel,
          'testTime': dateString,
          'closedEarly': 'yes'
        });
        saveData(subjID + "_incomplete_data.csv", jsPsych.data.get().csv());
      }
    },
    on_finish: function () {
      reached_finish = 1;
      jsPsych.data.addProperties({
        'subject': subjID,
        'assignedYesKey': jsPsych.pluginAPI.convertKeyCharacterToKeyCode(yeskey),
        'wordList': rand_stim_sel,
        'testTime': dateString,
        'closedEarly': 'no'
      });
      // write data file to server data folder
      saveData(subjID + "_data.csv", jsPsych.data.get().csv());
    },
    default_iti: 0
  });
```

However, if I close my browser prematurely after a certain point in the experiment, the data doesn't save. It doesn't save if I close the browser once reaching the last page of the repeat practice loop (code for last page and loop are shown below), or any time in the experiment after that point. Any idea what might be going wrong? I've also attached the whole experiment script here in case that helps. Thank you in advance for your thoughts on this. 

```
var practice_debrief = {
    type: 'html-button-response',
    choices: ['Read Instructions Again', 'Start Experiment'],
    stimulus: function () {
      var pracdata = jsPsych.data.get().filter({ 'sequence_type': 'practice' }).last(5);
      var accuracy = Math.floor(pracdata.filter({ correct: true }).count() / pracdata.count() * 100);

      var msg = "<p>You responded correctly <strong>" + accuracy + "%</strong> of " +
        "the time.</p>";

      if (accuracy > 85) {
        msg += "<p>Great job! Looks like you are ready to continue to the real thing.</p>"
      } else {
        msg += "<p>You might want to read through the instructions again to make sure that you understand what you need to do.</p>"
      }
      msg += "<p>If you'd like to read the instructions again, press the Read Instructions Again button. If you understand the instructions and are ready to start, press Start Experiment.</p>"

      return msg;
    },
  }
  all_practice.push(practice_debrief);

  var repeat_practice = {
    timeline: all_practice,
    loop_function: function (data) {
      return data.last(1).values()[0].button_pressed == 0;
    }
  }
  timeline.push(repeat_practice);
```

[lexical-decision.txt](https://github.com/jspsych/jsPsych/files/6065862/lexical-decision.txt)

**Comment by jodeleeuw** [comment, marked as answer]
Hi @patiencerosestevens,

Web browsers impose some limits on what you can and can't do during the unload event that closing the browser initiates. One limitation seems to be the amount of data that can be transferred. If you'd like to make sure to save all the data from a participant who exits prematurely I suggest saving the data after every trial using the `on_data_update` event handler in `jsPsych.init()`. 

[Docs here](https://www.jspsych.org/overview/callbacks/#on_data_update).

**Comment by patiencerosestevens** [comment]
@jodeleeuw Thank you so much for your quick and helpful reply, that fixed the issue!
